experimental/alp/lib/Transforms/legalize_vector.cpp

```diff
@@ -0,0 +1,211 @@
+
+    scf::ForOp newForOp = rewriter.create<scf::ForOp>(
+        forOp.getLoc(), forOp.lowerBound(), forOp.upperBound(), forOp.step(), newIterOperands);
+
+    Block &newBlock = newForOp.region().front();
+    SmallVector<Value, 4> extraYieldOperands;
+
+    oldBlock.walk([&](Operation *instr){
+        if (auto extractOp = dyn_cast<ExtractStridedSliceOp>(instr)){
+            // if offsets are the same
+            int64_t offset1 = extractFromI64ArrayAttr(extractOp.offsets())[0];
+            int64_t offset2 = extractFromI64ArrayAttr(extractOp.offsets())[1];
+            if (offset1 == working_offset1 && offset2 == working_offset2){
+                extractOp.getResult().replaceAllUsesWith(oldBlock.getArguments().back());
+                rewriter.eraseOp(extractOp);
+            }
+        } 
+        else if (auto insertOp = dyn_cast<InsertStridedSliceOp>(instr)){
+            int64_t offset1 = extractFromI64ArrayAttr(insertOp.offsets())[0];
+            int64_t offset2 = extractFromI64ArrayAttr(insertOp.offsets())[1];
+            if (offset1 == working_offset1 && offset2 == working_offset2){
+                insertOp.getResult().replaceAllUsesWith(insertOp.dest());
+                extraYieldOperands.push_back(insertOp.source());
+                rewriter.eraseOp(insertOp);
+            }
+       }
+    });
+
+    SmallVector<Value, 4> newBlockTransferArgs(newBlock.getArguments().begin(),
+                                              newBlock.getArguments().end());
+    rewriter.mergeBlocks(&oldBlock, &newBlock, newBlockTransferArgs);
+    auto clonedYieldOp = cast<scf::YieldOp>(newBlock.getTerminator());
+    SmallVector<Value> newYieldOperands = clonedYieldOp.getOperands();
+    for (auto val : extraYieldOperands){
+      newYieldOperands.push_back(val);
+    }
+    rewriter.setInsertionPoint(clonedYieldOp);
+    rewriter.create<scf::YieldOp>(newForOp.getLoc(), newYieldOperands);
+    rewriter.eraseOp(clonedYieldOp);
+
+    forOp.getResult(0).replaceAllUsesWith(newForOp.getResult(0));
+
+    SmallVector<Value, 4> newResults;
+    for (unsigned i =0; i<forOp.getNumResults(); i++){
+      newResults.push_back(newForOp.getResult(i));
+    }
+    rewriter.replaceOp(forOp, newResults);
+    // Get the users of the original result
+    auto users = newForOp.getResult(0).getUsers();
+
+    // And now let's change this
+    for (Operation *user : users){
+      if (auto extractOp = dyn_cast<ExtractStridedSliceOp>(user)){
+          int64_t offset1 = extractFromI64ArrayAttr(extractOp.offsets())[0];
+          int64_t offset2 = extractFromI64ArrayAttr(extractOp.offsets())[1];
+          if (offset1 == working_offset1 && offset2 == working_offset2){
+              extractOp.getResult().replaceAllUsesWith(newForOp.getResults().back());
+              rewriter.eraseOp(extractOp);
+          }
+      }
+    }
+    return success();
+  }
+};
+
+struct LegalizePass : public LegalizePassBase<LegalizePass>{
```

Comment:
I am not too sure what this pass does as there are not tests but I'd say it looks a bit fishy and at the intersection of other things that should exist in core.
Maybe if you describe the objective here I can provide some links?

Reply:
We discussed this here: https://llvm.discourse.group/t/outer-product-spilling/4405/7. I am basically chunking writes, reads and contracts in chunks of 4. As you correctly pointed out [here](https://llvm.discourse.group/t/outer-product-spilling/4405/3),  `scf.for + iter_args throw a wrench into the unrolling patterns`. I added the code to remove the `wrench`